### PR TITLE
Fix for 11614 - Adding HasPrecision() APIs

### LINKED ISSUE #11614: Add HasPrecision() Fluent API
Our type mapping supports precision and scale. We should add back EF6's `HasPrecision()` method on `PropertyBuilder`.

### PR DESCRIPTION
**Pull request by lajones**
Fixes #11614.

Added `HasPrecision(precision)` and `HasPrecision(precision, scale)` to `PropertyBuilder`.
Updates to scaffolding and migrations and to process all facets for all TypeMappings in Scaffolding.
Updated tests to match.

### CODE REVIEW

File: src/EFCore.Design/Scaffolding/Internal/ScaffoldingTypeMapper.cs [hunk 1]
```diff
@@ -57,112 +57,108 @@ public ScaffoldingTypeMapper([NotNull] IRelationalTypeMappingSource typeMappingS
...
-
-                    scaffoldMaxLength = sizedMapping.Size != byteArrayMapping.Size ? byteArrayMapping.Size : null;
-                }
-            }
-            else if (mapping.ClrType == typeof(string))
-            {
-                // Check for inference
-                var stringMapping = _typeMappingSource.FindMapping(
-                    typeof(string),
+                    fixedLength: false,
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Scaffolding/Internal/ScaffoldingTypeMapper.cs]: Should be `null`

File: src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs [hunk 2]
```diff
@@ -150,7 +150,7 @@ private static void FormatParameterValue(StringBuilder builder, object parameter
             {
                 builder
                     .Append('\'')
-                    .Append(((DateTime)parameterValue).ToString("s"))
+                    .Append(((DateTime)parameterValue).ToString("o"))
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs]: Why was this changed?

File: src/EFCore.Relational/Storage/RelationalTypeMappingSource.cs [hunk 3]
```diff
@@ -435,16 +435,30 @@ RelationalTypeMapping IRelationalTypeMappingSource.FindMapping(MemberInfo member
...
             }
 
             return storeTypeName;
         }
+
+        /// <summary>
+        /// Returns the store type name bases which interpret
+        /// nameBase(n) as a precision rather than a length
+        /// </summary>
+        protected virtual ICollection<string> StoreTypeNameBasesUsingPrecision
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Storage/RelationalTypeMappingSource.cs]: Would be more extensible if this returned a `bool` instead

File: src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeOffsetTypeMapping.cs [hunk 4]
```diff
@@ -16,7 +16,17 @@ namespace Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal
...
+        private readonly string[] _dateTimeOffsetFormats =
+        {
+            "'{0:yyyy-MM-ddTHH:mm:sszzz}'",
+            "'{0:yyyy-MM-ddTHH:mm:ss.fzzz}'",
+            "'{0:yyyy-MM-ddTHH:mm:ss.ffzzz}'",
+            "'{0:yyyy-MM-ddTHH:mm:ss.fffzzz}'",
+            "'{0:yyyy-MM-ddTHH:mm:ss.ffffzzz}'",
+            "'{0:yyyy-MM-ddTHH:mm:ss.fffffzzz}'",
+            "'{0:yyyy-MM-ddTHH:mm:ss.ffffffzzz}'",
+            "'{0:yyyy-MM-ddTHH:mm:ss.fffffffzzz}'"
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeOffsetTypeMapping.cs]: Add a comment stating that the order is significant

File: src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs (hunk 2 above)
Comment (lajones) [reply src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs]: We were only setting `DateTime` parameter values to the second. See the changes to e.g. `BuiltInDataTypesSqlServerTest` for examples. Without this it was impossible to see the fractions of seconds - which is what precision sets. Also note we already use this format for `DateTimeOffset` types.

File: src/EFCore.Relational/Storage/RelationalTypeMappingSource.cs (hunk 3 above)
Comment (lajones) [reply src/EFCore.Relational/Storage/RelationalTypeMappingSource.cs]: Done

File: src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeOffsetTypeMapping.cs (hunk 4 above)
Comment (lajones) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerDateTimeOffsetTypeMapping.cs]: Done

File: src/EFCore.Design/Scaffolding/Internal/ScaffoldingTypeMapper.cs (hunk 1 above)
Comment (lajones) [reply src/EFCore.Design/Scaffolding/Internal/ScaffoldingTypeMapper.cs]: Done

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs [hunk 5]
```diff
@@ -360,7 +360,9 @@ private RelationalTypeMapping FindRawMapping(RelationalTypeMappingInfo mappingIn
             return null;
         }
 
-        protected override ICollection<string> StoreTypeNameBasesUsingPrecision
-            => new List<string>() { "decimal", "dec", "numeric", "datetime2", "datetimeoffset" };
+        private readonly List<string> _nameBasesUsingPrecision =
```
Comment (lajones) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs]: Should be static. I'll fix.

Comment (lajones) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs]: Done

**Comment by lajones** [comment]
Note: many thanks to @SARAVANA1501, in particular his pull request #18237. I used that to get started on this.

**Comment by Tommigun1980** [comment]
Any chance of adding this as an attribute also?

Thanks.

**Comment by AndriySvyryd** [comment]
@Tommigun1980 See https://github.com/dotnet/efcore/issues/17914
